### PR TITLE
refactor: use cobra to implement the dfget client

### PR DESCRIPTION
**Pull request by fengzixu**
Signed-off-by: fengzixu <hnustphoenix@gmail.com>

Related to: https://github.com/dragonflyoss/Dragonfly/issues/249

### Ⅰ. Describe what this PR did

Using cobra to implement the dfget client

### Ⅱ. Does this pull request fix one issue?

https://github.com/dragonflyoss/Dragonfly/issues/249

### Ⅲ. Why don't you add test cases (unit test/integration test)? (你真的觉得不需要加测试吗？)
I will add ut before the sign "WIP" was removed.

### Ⅳ. Describe how to verify it

I run three command:

1. `./dfget -u 'https://raw.githubusercontent.com/dragonflyoss/Dragonfly/master/docs/images/logo/dragonfly-linear.png'    ` Downloading files successfully.
2. `./dfget version` Showing the version of dfget.
3.  `./dfget`  Showing the message about help.

### Ⅴ. Special notes for reviews

There are two problems I want to discuss with developers in the community:

1. When I use cobra to implement the dfget client in cmd/dfget, should I remove the /dfget/cli ? Or import /dfget/cli in cmd/dfget/root.go and use the format cli.xxx to use the function about dfget client.
2. Whether existing approaches which implemented the dfget client in this PR is OK ? If it is accessible, I will add ut for it and delete useless code in /dfget/cli

### CODE REVIEW

**Comment by fengzixu** [comment]
This PR perhaps delete some existed codes about dfget client. So, I submit a part of logic about modification. Please give me a few suggestions. I will find a good approach to refactor dfget client together you.

File: cmd/dfget/app/root.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
+package app
+
+import (
+	"fmt"
+
```
Comment (Starnop) [inline cmd/dfget/app/root.go]: delete this blank?

**Comment by fengzixu** [comment]
@Starnop Request review. Please check ` Special notes for reviews`

File: cmd/dfget/app/root.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	cfg "github.com/dragonflyoss/Dragonfly/dfget/config"
+	"github.com/dragonflyoss/Dragonfly/dfget/core"
+	"github.com/dragonflyoss/Dragonfly/dfget/errors"
+	"github.com/dragonflyoss/Dragonfly/dfget/util"
+	"github.com/sirupsen/logrus"
```
Comment (Starnop) [inline cmd/dfget/app/root.go]: Add a blank here?
 We suggest split the imports into three group with a blank:
1. the packages belong to go standard library
2. the packages belong to our project 
3. the packages belong to Third-party

File: cmd/dfget/app/root.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+		"show progress bar, it's conflict with '--console'")
+	rootCmd.PersistentFlags().BoolVar(&cfg.Ctx.Console, "console", false,
+		"show log on console, it's conflict with '--showbar'")
+	rootCmd.PersistentFlags().BoolVar(&cfg.Ctx.Verbose, "verbose", false,
+		"be verbose")
+	rootCmd.PersistentFlags().BoolVarP(&cfg.Ctx.Help, "help", "h", false,
+		"show help information")
+
+	rootCmd.PersistentFlags().MarkDeprecated("exceed", "please use '--timeout' or '-e' instead")
+	// be compatible with dfget python version
```
Comment (Starnop) [inline cmd/dfget/app/root.go]: Could please help get it out of the function of `initFlags`, It's wired here. ^_^

File: cmd/dfget/app/root.go (hunk 2 above)
Comment (fengzixu) [reply cmd/dfget/app/root.go]: Sure. I firstly use goland, it cannot help me to format import. I will revise it quickly.

**Comment by Starnop** [comment]
> There are two problems I want to discuss with developers in the community:

When I use cobra to implement the dfget client in cmd/dfget, should I remove the /dfget/cli ? Or import /dfget/cli in cmd/dfget/root.go and use the format cli.xxx to use the function about dfget client.
Whether existing approaches which implemented the dfget client in this PR is OK ? If it is accessible, I will add ut for it and delete useless code in /dfget/cli

In my opinion, one file is enough, maybe `/dfget/cli ` is not necessary we completely refactor. WDYT?

File: cmd/dfget/app/root.go (hunk 3 above)
Comment (fengzixu) [reply cmd/dfget/app/root.go]: Remove or Put it into an other place?

File: vendor/vendor.json [hunk 4]
```diff
@@ -315,5 +333,5 @@
 			"revisionTime": "2018-03-28T19:50:20Z"
 		}
 	],
-	"rootPath": "github.com/dragonflyoss/Dragonfly"
+	"rootPath": "github.com/dragonflyoss/dragonfly"
```
Comment (Starnop) [inline vendor/vendor.json]: Why change it ? Our root path should be `github.com/dragonflyoss/Dragonfly` ,not `github.com/dragonflyoss/dragonfly`.

Comment (fengzixu) [reply vendor/vendor.json]: OK. I roll back it.

File: cmd/dfget/app/root.go [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+		cfg.Ctx.TotalLimit = cfg.Props.TotalLimit
+	}
+
+	if cfg.Ctx.ClientQueueSize == 0 {
+		cfg.Ctx.ClientQueueSize = cfg.Props.ClientQueueSize
+	}
+
+	cfg.Ctx.Filter = transFilter(filter)
+
+	// be compatible with dfget python version
```
Comment (fengzixu) [inline cmd/dfget/app/root.go]: Is there necessary to compatible with python version? @Starnop

Comment (Starnop) [reply cmd/dfget/app/root.go]: @lowzj  Can we make a unity here? In my opinion, a string type might be more friendly for users. WDYT?

File: cmd/dfget/app/root.go [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+		os.Exit(0)
+	},
+}
+
+func init() {
+	initFlags()
+}
+
+func checkParameters() {
+	fmt.Println(os.Args)
```
Comment (Starnop) [inline cmd/dfget/app/root.go]: delete it or make it less arbitrary.

File: cmd/dfget/app/root.go [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+	if len(os.Args) < 3 {
```
Comment (Starnop) [inline cmd/dfget/app/root.go]: Just out of curiosity, why change `2` to `3`?

File: cmd/dfget/app/version.go [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+import (
+	"fmt"
+
+	"github.com/dragonflyoss/Dragonfly/version"
+
+	"github.com/spf13/cobra"
+)
+
+var versionCmd = &cobra.Command{
```
Comment (Starnop) [inline cmd/dfget/app/version.go]: Maybe provide our own help command? WDYT?

File: cmd/dfget/app/root.go (hunk 6 above)
Comment (fengzixu) [reply cmd/dfget/app/root.go]: Sorry. I add some code about testing.

File: cmd/dfget/app/root.go (hunk 7 above)
Comment (fengzixu) [reply cmd/dfget/app/root.go]: When I type the commands like this `dfget -u`, the `len(os.Args)` is 2. That is illegal input. For dfget, We must supply a valid URL.

File: cmd/dfget/app/version.go (hunk 8 above)
Comment (fengzixu) [reply cmd/dfget/app/version.go]: I cannot understand your mean. Is that info about version is unnecessary?

File: cmd/dfget/app/root.go [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+	localLimit string
+	totalLimit string
+	filter     string
+)
+
+var rootCmd = &cobra.Command{
+	Use:   "dfget",
+	Short: "The dfget is the client of Dragonfly.",
+	Long:  "The dfget is the client of Dragonfly, a non-interactive P2P downloader.",
+	Run: func(cmd *cobra.Command, args []string) {
```
[the remark this answers is not in the document]
Comment (fengzixu) [reply cmd/dfget/app/root.go]: Please describe your mean clearly.

File: cmd/dfget/app/root.go (hunk 7 above)
Comment (Starnop) [reply cmd/dfget/app/root.go]: However, there are also parameters of bool.

File: cmd/dfget/app/version.go (hunk 8 above)
Comment (Starnop) [reply cmd/dfget/app/version.go]: It is ok at present, but we can also do some custom output like [this](https://github.com/dragonflyoss/Dragonfly/blob/master/dfget/cli/cli_parameters.go#L106) to make it more friendly, WDYT?

File: cmd/dfget/app/root.go (hunk 7 above)
Comment (fengzixu) [reply cmd/dfget/app/root.go]: Please give me an example which can prove `len(os.Args) < 2` is right

Comment (fengzixu) [reply cmd/dfget/app/root.go]: I see. I will modify it.

File: cmd/dfget/app/root.go [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+		cfg.Ctx.ClientLogger.Infof("cmd params:%q", args)
+		initProperties()
+		cfg.Ctx.ClientLogger.Infof("context:%s", cfg.Ctx)
+		err := core.Start(cfg.Ctx)
+		if err != nil {
+			util.Printer.Println(resultMsg(cfg.Ctx, time.Now(), err))
+			os.Exit(err.Code)
+		}
+		util.Printer.Println(resultMsg(cfg.Ctx, time.Now(), err))
+		os.Exit(0)
```
Comment (Starnop) [inline cmd/dfget/app/root.go]: Extract the above code as a separate function. It's just my personal advice. 😄

File: cmd/dfget/app/root.go (hunk 5 above)
Comment (fengzixu) [reply cmd/dfget/app/root.go]: Sure

File: cmd/dfget/app/root.go (hunk 10 above)
Comment (fengzixu) [reply cmd/dfget/app/root.go]: Your mean: separate `util.Printer.Println(resultMsg(cfg.Ctx, time.Now(), err))` to 
```go
resultMsg := resultMsg(cfg.Ctx, time.Now(), err)
util.Printer.Println(resultMsg)
```

**Comment by fengzixu** [comment]
For the `help` command, I think that we don't need to add extra help information about dfget at now. The default `help` command has already satisfied our requirements. In the near feature, If we want to expand it, we can add it at that time. @Starnop

**Comment by Starnop** [comment]
LGTM.

**Comment by fengzixu** [comment]
Can we move forward this PR ? @Starnop @lowzj

**Comment by allencloud** [comment]
LGTM

**Pull request merged**